File: app/src/main/java/org/adaway/ui/log/LogViewModel.java
package org.adaway.ui.log;

import android.app.Application;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import org.adaway.AdAwayApplication;
import org.adaway.db.AppDatabase;
import org.adaway.db.dao.HostEntryDao;
import org.adaway.db.dao.HostListItemDao;
import org.adaway.db.entity.HostListItem;
import org.adaway.db.entity.ListType;
import org.adaway.model.adblocking.AdBlockMethod;
import org.adaway.model.adblocking.AdBlockModel;
import org.adaway.util.AppExecutors;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;

import static org.adaway.db.entity.HostsSource.USER_SOURCE_ID;

/**
 * This class is an {@link AndroidViewModel} for the {@link LogActivity}.
 *
 * @author Bruce BUJON (bruce.bujon(at)gmail(dot)com)
 */
public class LogViewModel extends AndroidViewModel {
    private final AdBlockModel adBlockModel;
    private final HostListItemDao hostListItemDao;
    private final HostEntryDao hostEntryDao;
    private final MutableLiveData<List<LogEntry>> logEntries;
    private final MutableLiveData<Boolean> recording;
    private LogEntrySort sort;

    public LogViewModel(@NonNull Application application) {
        super(application);
        this.adBlockModel = ((AdAwayApplication) application).getAdBlockModel();
        this.hostListItemDao = AppDatabase.getInstance(application).hostsListItemDao();
        this.hostEntryDao = AppDatabase.getInstance(application).hostEntryDao();
        this.logEntries = new MutableLiveData<>();
        this.recording = new MutableLiveData<>(this.adBlockModel.isRecordingLogs());
        this.sort = LogEntrySort.TOP_LEVEL_DOMAIN;
    }

    public boolean areBlockedRequestsIgnored() {
        return this.adBlockModel.getMethod() == AdBlockMethod.ROOT;
    }

    public LiveData<List<LogEntry>> getLogs() {
        return this.logEntries;
    }

    public void clearLogs() {
        this.adBlockModel.clearLogs();
        this.logEntries.postValue(Collections.emptyList());
    }

    public void updateLogs() {
        AppExecutors.getInstance().diskIO().execute(
                () -> {
                    // Get tcpdump logs
                    List<LogEntry> logItems = this.adBlockModel.getLogs()
                            .parallelStream()
                            .map(log -> {
                                ListType type = this.hostEntryDao.getTypeOfHost(log);
                                return new LogEntry(log, type);
                            })
                            .sorted(this.sort.comparator())
                            .collect(Collectors.toList());
                    // Post result
                    this.logEntries.postValue(logItems);
                }
        );
    }

    public void toggleSort() {
        this.sortDnsRequests(this.sort == LogEntrySort.ALPHABETICAL ?
                LogEntrySort.TOP_LEVEL_DOMAIN :
                LogEntrySort.ALPHABETICAL
        );
    }

    public LiveData<Boolean> isRecording() {
        return this.recording;
    }

    public void toggleRecording() {
        boolean recording = !this.adBlockModel.isRecordingLogs();
        this.adBlockModel.setRecordingLogs(recording);
        this.recording.postValue(recording);
    }

    public void addListItem(@NonNull String host, @NonNull ListType type, String redirection) {
        // Create new host list item
        HostListItem item = new HostListItem();
        item.setType(type);
        item.setHost(host);
        item.setRedirection(redirection);
        item.setEnabled(true);
        item.setSourceId(USER_SOURCE_ID);
        // Insert host list item
        AppExecutors.getInstance().diskIO().execute(() -> this.hostListItemDao.insert(item));
        // Update log entries
        updateLogEntryType(host, type);
    }

    public void removeListItem(@NonNull String host) {
        // Delete host list item
        AppExecutors.getInstance().diskIO().execute(() -> this.hostListItemDao.deleteUserFromHost(host));
        // Update log entries
        updateLogEntryType(host, null);
    }

    private void updateLogEntryType(@NonNull String host, ListType type) {
        // Get current values
        List<LogEntry> entries = this.logEntries.getValue();
        if (entries == null) {
            return;
        }
        // Update entry type
        List<LogEntry> updatedEntries = entries.stream()
                .map(entry -> entry.getHost().equals(host) ? new LogEntry(host, type) : entry)
                .collect(Collectors.toList());
        // Post new values
        this.logEntries.postValue(updatedEntries);
    }

    private void sortDnsRequests(LogEntrySort sort) {
        // Save current sort
        this.sort = sort;
        // Apply sort to values
        List<LogEntry> entries = this.logEntries.getValue();
        if (entries != null) {
            List<LogEntry> sortedEntries = new ArrayList<>(entries);
            sortedEntries.sort(this.sort.comparator());
            this.logEntries.postValue(sortedEntries);
        }
        // Notify user
        Toast.makeText(
                getApplication(),
                this.sort.getName(),
                Toast.LENGTH_SHORT
        ).show();
    }
}

File: app/src/main/java/org/adaway/ui/log/LogViewModel.kt
package org.adaway.ui.log

import android.app.Application
import android.widget.Toast
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import org.adaway.AdAwayApplication
import org.adaway.db.AppDatabase
import org.adaway.db.dao.HostEntryDao
import org.adaway.db.dao.HostListItemDao
import org.adaway.db.entity.HostListItem
import org.adaway.db.entity.HostsSource.USER_SOURCE_ID
import org.adaway.db.entity.ListType
import org.adaway.model.adblocking.AdBlockMethod
import org.adaway.model.adblocking.AdBlockModel
import org.adaway.util.ExpressiveToast

class LogViewModel(application: Application) : AndroidViewModel(application) {
    private val adBlockModel: AdBlockModel = (application as AdAwayApplication).adBlockModel
    private val hostListItemDao: HostListItemDao = AppDatabase.getInstance(application).hostsListItemDao()
    private val hostEntryDao: HostEntryDao = AppDatabase.getInstance(application).hostEntryDao()
    private var sort = LogEntrySort.TOP_LEVEL_DOMAIN

    private val _logs = MutableStateFlow<List<LogEntry>>(emptyList())
    val logs: StateFlow<List<LogEntry>> = _logs

    private val _recording = MutableStateFlow(adBlockModel.isRecordingLogs)
    val recording: StateFlow<Boolean> = _recording

    private val _refreshing = MutableStateFlow(false)
    val refreshing: StateFlow<Boolean> = _refreshing

    fun areBlockedRequestsIgnored(): Boolean = adBlockModel.method == AdBlockMethod.ROOT

    fun clearLogs() {
        adBlockModel.clearLogs()
        _logs.value = emptyList()
        _refreshing.value = false
    }

    fun updateLogs() {
        viewModelScope.launch {
            _refreshing.value = true
            val logItems = withContext(Dispatchers.IO) {
                adBlockModel.logs
                    .parallelStream()
                    .map { log -> LogEntry(log, hostEntryDao.getTypeOfHost(log)) }
                    .sorted(sort.comparator())
                    .toList()
            }
            _logs.value = logItems
            _refreshing.value = false
        }
    }

    fun toggleSort() {
        sortDnsRequests(
            if (sort == LogEntrySort.ALPHABETICAL) {
                LogEntrySort.TOP_LEVEL_DOMAIN
            } else {
                LogEntrySort.ALPHABETICAL
            }
        )
    }

    fun toggleRecording() {
        val recording = !adBlockModel.isRecordingLogs
        adBlockModel.setRecordingLogs(recording)
        _recording.value = recording
    }

    fun addListItem(host: String, type: ListType, redirection: String?) {
        val item = HostListItem().apply {
            this.type = type
            this.host = host
            this.redirection = redirection
            isEnabled = true
            sourceId = USER_SOURCE_ID
        }
        viewModelScope.launch(Dispatchers.IO) {
            hostListItemDao.insert(item)
        }
        updateLogEntryType(host, type)
    }

    fun removeListItem(host: String) {
        viewModelScope.launch(Dispatchers.IO) {
            hostListItemDao.deleteUserFromHost(host)
        }
        updateLogEntryType(host, null)
    }

    private fun updateLogEntryType(host: String, type: ListType?) {
        _logs.value = _logs.value.map { entry ->
            if (entry.host == host) LogEntry(host, type) else entry
        }
    }

    private fun sortDnsRequests(sort: LogEntrySort) {
        this.sort = sort
        _logs.value = _logs.value.sortedWith(sort.comparator())
        ExpressiveToast.makeText(
            getApplication(),
            sort.getName(),
            Toast.LENGTH_SHORT
        ).show()
    }
}
